### PR TITLE
[C10D] Make new_group eager when used with comm_split

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129284

If users pass `device_id` to init_process_group, they enable eager init
for the default group.  Then if they subsequently call `new_group`, the
device_id argument is not required as it should be assumed to match the
one used for init_process_group.

However, both `init_process_group` and `new_group` apis share a helper
function, which expects a `device_id` value that defaults to None.  When
it's None, eager initialization is disabled.

This PR ensures that if a device_id was passed to init_process_group,
the same device_id will automatically be fed into the helper function
for any new_group calls that follow.

**Test plan**
I found an existing test in CI  `test_comm_split_subgroup` that failed after my change, because it was asserting that backend comm_split counter did not increment eagerly, and its behavior had changed to increment eagerly.  I updated the test in the PR to pass with my change.

I also tested locally via simple program with TORCH_CPP_LOG_LEVEL=INFO and
observed eager initialization of the 'lows' and 'highs' PGs before the
'Here' print.

```
import torch
import torch.distributed as dist
dist.init_process_group(backend="nccl", device_id =torch.device(f"cuda:{torch.distributed.get_node_local_rank(0)}"))
dist.new_group([0, 1], group_desc="lows")
dist.new_group([2, 3], group_desc="highs")
print("Here")
torch.distributed.destroy_process_group()
```

Output:
https://gist.github.com/wconstab/88a5ba0b970244ca1f79133f989e0349

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225 @chauhang @d4l3k

### CODE REVIEW

**Review by pavanbalaji — APPROVED** [review]
LGTM!

**Comment by wconstab** [comment]
This failure was actually an existing unit test needing to be updated to match the new behavior.  i pushed the test update and updated the PR desc accordingly.

**Comment by wconstab** [comment]
@eqy @nvcastet pls also take a look at this one, hopefully no issue here

**Review by fduwjj — APPROVED** [review]
LGTM.

**Review by d4l3k — APPROVED** [review]
LGTM

**Review by nvcastet — APPROVED** [review]
Looks good!

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -4394,6 +4394,7 @@ def _new_group_with_tag(
     global _world
 
     default_pg = _get_default_group()
+    device_id = default_pg.bound_device_id or None
```
Comment (nvcastet) [inline torch/distributed/distributed_c10d.py:4397]: is it really necessary to have `or None`? `bound_device_id` property will already return `None` when unset

Comment (wconstab) [reply torch/distributed/distributed_c10d.py:4397]: Oops, yea this is kinda dumb. I will submit a cleanup PR.

**Pull request closed without merging**